R/redist_constr.R

```diff
@@ -655,10 +721,110 @@ add_constr_custom <- function(constr, strength, fn) {
 
     rlang::fn_env(fn) <- constr_env
 
-    new_constr <- list(strength = strength, fn = fn)
+    new_constr <- list(strength = strength, score_districts_only=score_districts_only,
+                       fn = fn)
     add_to_constr(constr, "custom", new_constr)
 }
 
+#' @param fn A function
+#' @rdname constraints
+#' @export
+add_constr_custom_plan <- function(constr, strength, fn) {
+    if (!inherits(constr, "redist_constr")) cli_abort("Not a {.cls redist_constr} object")
+    if (strength <= 0) cli_warn("Nonpositive strength may lead to unexpected results")
+
+    args <- rlang::fn_fmls(fn)
+    if (length(args) != 2) cli_abort("Function must take exactly two arguments.")
+
+    constr_env = rlang::fn_env(fn)
+    constr_env <- rlang::env(constr_env)
+    # every symbol used in the function (except the 2 arguments)
+    var_names = setdiff(
+        all.names(rlang::fn_body(fn)),
+        names(args)
+    )
+
+    for (nm in var_names) {
+        found = find_env(nm, constr_env)
+        if (!is.null(found) &&
+            !identical(found, rlang::base_env()) &&
+            !identical(found, constr_env) &&
+            !identical(found, rlang::pkg_env("redist"))) {
+            constr_env[[nm]] = get(nm, envir=found)
+        }
+    }
+
+    if (!is.null(plan <- get_existing(attr(constr, "data")))) {
+        fake_sizes <- rep(1L, length(unique(plan)))
+        out <- tryCatch(fn(plan, fake_sizes), error = function(e) {
+            cli_abort(c("Ran into an error testing custom constraint
+                        on the existing plan:",
+                        "x" = e$message))
+        })
+        if (!is.numeric(out) || length(out) != 1 || !is.finite(out))
+            cli_abort(c("Evaluting custom constraint on the existing plan failed.",
+                        "*" = "The constraint function should return a single scalar value.",
+                        "*" = "Make sure that your constraint function tests all edge cases
+                             and never returns {.val {NA}} or {.val {Inf}}."))
+    }
+
+    rlang::fn_env(fn) <- constr_env
+
+    new_constr <- list(strength = strength,
+                       fn = fn)
+    add_to_constr(constr, "custom_plan", new_constr)
+}
+
+
+#' @param fn A function
+#' @rdname constraints
+#' @export
+add_hard_constr_custom_plan <- function(constr, fn) {
```

Comment:
See Slack discussion. Don't think we should separate pre- vs post- constraints. Instead, add a `thresh=Inf` arg